### DISCUSSION TITLE
Welcome to rlai Discussions!

### DISCUSSION THREAD

**Discussion by MatthewGerber** [MatthewGerber/rlai#5] [General]
## 👋 Welcome!
  Please feel free to
  * ask questions you’re wondering about;
  * share ideas; and
  * engage with others.

**Comment by zevla** [comment]
Hello, I want to know how to run a trained agent in rlai? You have some agents in .pickle format. But how do I use them? I tried everything. I also went through the github website for help. Nothing relevant there as well.

**Reply by MatthewGerber** [reply]
Hello! See [here](https://matthewgerber.github.io/rlai/#installation-and-use) for how to clone/install. I added an example of how to run a trained agent [here](https://matthewgerber.github.io/rlai/case_studies/mountain_car_continuous.html#results), for the continuous mountain car environment. I just pushed a fix that lets this agent run, so pull the latest code if needed.

**Reply by zevla** [reply]
I have followed your instructions and installed poetry and the dependencies. I have also been able to run a trained agent for the continuous mountain car environment. But when I try to change the environment for the swimmer, it doesnt work for some reason.

So this code works fine:

`rlai run --agent ./trained_agents/mountaincar_continuous/mountaincar_continuous_agent.pickle --environment rlai.core.environments.gymnasium.Gym --steps-per-second 35 --gym-id MountainCarContinuous-v0 --render-every-nth-episode 1 --n-runs 1 --T 2000 --plot --plot-environment`

But this doesn't:

`rlai run --agent ./trained_agents/swimmer/swimmer-agent.pickle --environment rlai.core.environments.gymnasium.Gym --steps-per-second 35 --gym-id Swimmer-v5 --render-every-nth-episode 1 --n-runs 1 --T 2000 --plot --plot-environment`


This is the error I get:

> C:\Users\X\Desktop\backup\rlai_github\rlai\src\rlai\runners\agent_in_environment.py:42: UserWarning: No random seed provided to the trainer. Results will not be replicable. Consider passing --random-seed argument.
>   warnings.warn(
> C:\Users\X\Desktop\backup\rlai_github\rlai\src\rlai\core\environments\gymnasium.py:606: UserWarning: No Gym customizer is available, so the state-dimension names for are unknown. Defaulting to numbers.
>   warnings.warn(
> Traceback (most recent call last):
>   File "C:\Users\X\AppData\Local\pypoetry\Cache\virtualenvs\rlai-ZkqGkj4s-py3.11\Scripts\\rlai", line 6, in <module>
>     sys.exit(run())
>              ^^^^^
>   File "C:\Users\X\Desktop\backup\rlai_github\rlai\src\rlai\runners\top_level.py", line 38, in run
>     return parsed_args.func(unparsed_args)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\X\Desktop\backup\rlai_github\rlai\src\rlai\runners\agent_in_environment.py", line 60, in run
>     agents = [pickle.load(f)]
>               ^^^^^^^^^^^^^^
> ModuleNotFoundError: No module named 'rlai.agents'
> C:\Users\X\AppData\Local\pypoetry\Cache\virtualenvs\rlai-ZkqGkj4s-py3.11\Lib\site-packages\gymnasium\envs\mujoco\mujoco_rendering.py:403: UserWarning: WARN: Environment was not properly closed using 'env.close()'. Please ensure to close the environment explicitly. GLFW module or dependencies are unloaded. Window cleanup might not have completed.

I have looked into the relevant python codes in the folder (_top_level.py_ and _agent_in_environment.py_) but I cant figure out the issue.

**Reply by MatthewGerber** [reply]
It looks like that `swimmer-agent.pickle` was created before a code refactor that changed the package structure. You would need to train a new swimmer agent using a command like the following:

```
rlai train --random-seed 12345 --agent rlai.policy_gradient.ParameterizedMdpAgent --gamma 0.99 --environment rlai.core.environments.gymnasium.Gym --gym-id Swimmer-v5 --render-every-nth-episode 100 --video-directory ~/Desktop/swimmer_videos --T 500 --train-function rlai.policy_gradient.monte_carlo.reinforce.improve --num-episodes 4000 --v-S rlai.state_value.function_approximation.ApproximateStateValueEstimator --feature-extractor rlai.core.environments.gymnasium.SignedCodingFeatureExtractor --function-approximation-model rlai.models.sklearn.SKLearnSGD --loss squared_error --sgd-alpha 0.0 --learning-rate constant --eta0 0.00001 --policy rlai.policy_gradient.policies.continuous_action.ContinuousActionBetaDistributionPolicy --policy-feature-extractor rlai.core.environments.gymnasium.SignedCodingFeatureExtractor --alpha 0.00001 --update-upon-every-visit True --num-episodes-per-checkpoint 1000 --checkpoint-path ~/Desktop/swimmer_checkpoint.pickle --save-agent-path ~/Desktop/swimmer_agent.pickle --log INFO --num-episodes-per-policy-update-plot 100
```
